### PR TITLE
[#3325][feat] Add MCTS and TOT tree-based inference controllers to Scaffolding

### PR DESCRIPTION
**Pull request by therealnaveenkamal**
## Summary by CodeRabbit

- New Features
  - Added tree-based inference controllers: Monte Carlo Tree Search (MCTS) and Tree of Thoughts (ToT) for structured reasoning with configurable depth, iterations, and thoughts-per-step, with optional reward scoring.
  - Public API now exposes PRM, MCTS, and ToT controllers.
  - Introduced CLI examples to run MCTS/ToT workflows with adjustable parameters.

- Documentation
  - New guide detailing MCTS and ToT methods, configuration options, and usage examples, including tips for tuning outputs.

## Description
This PR implements tree-based inference methods in Scaffolding to address this issue #3325 by adding:

- MCTSController: UCB1 selection, expansion via generation, simulation via PRM or heuristic, backpropagation; compact “Problem + Steps” final reasoning.
- TOTController: level-wise thought expansion with evaluation; supports PRM-based scoring; robust parsing of “Approach/Step/numbered/bulleted” formats; selection strategies (best, vote, random) with configurable branch_factor; final synthesis from best path.

Examples: `examples/scaffolding/run_mcts_example.py` and `examples/scaffolding/run_tot_example.py`.
Docs: concise `tensorrt_llm/scaffolding/README_TREE_METHODS.md`.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Comment by therealnaveenkamal** [comment]
cc @WeiHaocheng

File: tensorrt_llm/scaffolding/tree_controllers.py [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+
+            # Expansion
+            if not node.is_terminal and node.depth < self.max_depth:
+                prompt = self._create_expansion_prompt(node, goal)
+                gen_task = GenerationTask.create_from_prompt(prompt)
+                gen_task.max_tokens = 200
+                gen_task.temperature = 0.7
+
+                # Delegate to generation controller
+                yield from self.generation_controller.process([gen_task])
```
Comment (WeiHaocheng) [inline tensorrt_llm/scaffolding/tree_controllers.py:135]: For performance considerations, I suggest using “ParallelProcess” here to enable the worker to handle self.max_iterations of the gerneration request in parallel. The usage of “ParallelProcess” can be referred to MajorityVoteController in controller.py. The code here is link:
`
generation_controllers = []
tasks_list = []
generation_kwargs_list = []

for iteration in range(self.max_iterations):
     # Some code here
     generation_controllers.append(...)
     tasks_list.append(...)
     generation_kwargs_list.append(...)
     yield ParallelProcess(generation_controllers, tasks_list,
                              generation_kwargs_list)
   
`

The code for reward is similar here.

File: examples/scaffolding/run_mcts_example.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    # Initialize reward worker if provided
+    reward_controller = None
+    reward_worker = TRTLLMWorker.init_with_new_llm(
+        args.reward_model_dir,
+        backend="pytorch",
+        max_batch_size=4,
+        max_num_tokens=8192,
+        kv_cache_free_gpu_memory_fraction=0.2,
+        disable_overlap_scheduler=True)
```
Comment (WeiHaocheng) [inline examples/scaffolding/run_mcts_example.py:46]: Why do we need to set kv_cache_free_gpu_memory_fraction and disable_overlap_scheduler here?
Is any bug with overlap_scheduler?
Can we put kv_cache_free_gpu_memory_fraction into the argument?

File: examples/scaffolding/run_mcts_example.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import argparse
+
```
Comment (WeiHaocheng) [inline examples/scaffolding/run_mcts_example.py:2]: Have you run 
`pre-commit run --all-files`
for this commit?

File: examples/scaffolding/run_mcts_example.py (hunk 2 above)
Comment (therealnaveenkamal) [reply examples/scaffolding/run_mcts_example.py:46]: - kv_cache_free_gpu_memory_fraction: I used only one GPU between the generation model and the PRM. Setting a smaller fraction for gen (e.g., 0.1) and a slightly larger fraction for reward (e.g., 0.2) reduced OOMs. It also makes latency more stable when PRM scoring spikes briefly.

- disable_overlap_scheduler (for reward): The score in PRM sometime spikes...so overlap gives limited throughput upside but increases scheduling pressure. Disabling overlap on the reward side made co-scheduling with gen smoother on a single GPU. I think in case of multiple GPUs, this shouldn't be an issue.

Comment (therealnaveenkamal) [reply examples/scaffolding/run_mcts_example.py:46]: I've added them in the args

**Comment by therealnaveenkamal** [comment]
@WeiHaocheng i've made the changes. please review and let me know

File: tensorrt_llm/scaffolding/tree_controllers.py [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+                   key=lambda child: child.ucb1_score(exploration_constant))
+
+
+@dataclass
+class TOTNode(TreeNode):
+    """Node for Tree of Thoughts."""
+    thought: str = ""
+    confidence: float = 0.0
+    reasoning: str = ""
+    evaluation_score: float = 0.0
```
[the remark this answers is not in the document]
Comment (WeiHaocheng) [reply tensorrt_llm/scaffolding/tree_controllers.py]: I think there's no problem at all with having our own implementation adjustments. We just need to emphasize in the README that this idea comes from that paper, but the implementation details are not exactly the same.

File: examples/scaffolding/contrib/MCTS/run_mcts_example.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
```
Comment (WeiHaocheng) [inline examples/scaffolding/contrib/MCTS/run_mcts_example.py:1]: I suggest to keep the same directory name with the code in tensorrt_llm/scaffolding. This helps others find your example.

Comment (therealnaveenkamal) [reply examples/scaffolding/contrib/MCTS/run_mcts_example.py:1]: quick clarification - should I move both `run_mcts_example.py` and `run_tot_example.py` to `tensorrt_llm/scaffolding/TreeInference`?

**Review by WeiHaocheng — APPROVED** [review]

File: examples/scaffolding/contrib/MCTS/run_mcts_example.py (hunk 5 above)
Comment (WeiHaocheng) [reply examples/scaffolding/contrib/MCTS/run_mcts_example.py]: Sorry, maybe I make something confusing here. "run_mcts_example.py" and "run_tot_example.py" should be on "example". I mean you can put them at the "examples/scaffolding/contrib/TreeInference".

**Comment by therealnaveenkamal** [comment]
@WeiHaocheng there was a small lint error my side, fixed it. thanks.

**Review by QiJune — APPROVED** [review]
